packages/cli/src/baseCommand.js

```diff
@@ -0,0 +1,58 @@
+const { Command } = require('@oclif/command');
+const { isEmpty } = require('lodash');
+const { cli: ux } = require('cli-ux');
+
+const { readCredentials, checkCredentials } = require('./utils/api');
+
+class ZapierBaseCommand extends Command {
+  _version() {
+    return 'blah blah ok'; // this should overwrite the version string, but doesn't?
+  }
+
+  log(message) {
+    if (!['json', 'raw'].includes(this.flags.format)) {
+      super.log(message || '');
+    }
+  }
+
+  get prompt() {
+    return ux.prompt;
+  }
+  // AUTH
+  /**
+   * ensures credential file exists and looks correct. Doesn't guarantee creds are still valid, but the API call will blow up if that's the case. Cheap way to test early that user has a session
+   */
+  async probablyLoggedIn() {
+    const creds = readCredentials(false);
+    if (isEmpty(creds)) {
+      this.log('Before you can do that, you need to be logged in.\n');
+      // run login command, which we can't call from here since it'll eventually import this?
+      this.login();
```

Comment:
this was one of the issues I ran into. In our currently CLI, we've got a [special file](https://github.com/zapier/zapier-platform/blob/e8d4fc588bdb2a15059a151057b607a8057b0a06/packages/cli/src/utils/command-invocations.js) that imports the `login` and `register` commands and can run them as functions. this lets us drop into the login flow in another command, so we interrupt the dev less (instead of throwing an error and asking them to run `zapier login`, it just prompts them when they've run an authed command).

So, we've got two major options:

1. move all the login code into a standalone function. It's easily re-useable, but won't be able to take advantage of any of the methods we add onto `BaseCommand` since it's a standalone. We could pass the relevant methods in, but that'll make it really annoying to call (eg `login(this.log, this.x, this.y)`)
2. make exceptions for `login` and `register` and build those full methods onto the base command. Then it's easy to call them in any other command, but seems like a bit of a code smell. They have a doc about [command code reuse](https://oclif.io/docs/running_programmatically#sharing-code-with-modules), but it's a little confusing. I've opened [an issue](https://github.com/oclif/oclif/issues/274) to get more info. Maybe we can eat our cake and have it too!

Reply:
It's [definitely doable](https://oclif.io/docs/running_programmatically), but those docs include this:

> First, it is generally a bad idea to run a command directly as the command exports a user interface, not a code interface. It's a design smell that should rarely (if ever) be used. Generally speaking, it's better to break up the code so that it can be called directly rather than as a command. We'll show this better method first.

That (as I read it) encourages us to move it into a standalone function. But, then it's harder to use utils (namely logging) that are used elsewhere.

I think if I had to pick one, I would just run the other commands directly (like it says not to). I can't think of a cleaner way to do do it. ¯\\\_(ツ)_/¯